[KPT CODELAB] (2/3) kpt  config validation (#5026)
GoogleContainerTools/skaffold: codelab/01_kpt-validate/tutorial.md

# Validate Your Configurations in skaffold Workflow

<walkthrough-disable-features toc></walkthrough-disable-features>

## Introduction

#### What is kpt?

Kpt is [an OSS tool](https://github.com/GoogleContainerTools/kpt) for Kubernetes packaging, which uses a standard format to bundle, publish, customize, update, and apply configuration manifests.

#### What kpt can help you
    
*  You will get an hands-on off-the-shelf experience about the **GitOps** CI/CD workflow in skaffold.
*  You can validate each of your config changes **declaratively**.
*  You **won't** encounter **version conflict** if the config hydration (a.k.a kustomize) mismatch with the deployment tool (e.g. kubectl). 
*  You can prune your resources accurately with [a three-way merge strategy](https://kubectl.docs.kubernetes.io/pages/app_management/field_merge_semantics.html). 

#### What you'll learn
    
*  How to add a validation to your config changes. 
*  How to define validation rules in the form of a declarative configuration.
*  How to use the validation in kustomized resources. 

> 

Reviewer: The content here isn't consistent with the bullet points in `README.md`. Is this intentional?

Author: Yeah, I'm planning to merge the validate-1 and deploy-2 README.md to a single file. But since they are currently two separate PRs. I'd prefer not making the change. You can ignore README.md since it would be changed.

Reviewer: works for me 👍🏼

Or just follow this codelab, we will explain what happens in each step.

#### Install `skaffold`

*   Check if `skaffold` is installed and the installed version is >= v1.17.0
    ```bash
    skaffold version
    ```

*   If you haven't installed `skaffold` previously, run
    ```bash
    curl -Lo skaffold https://storage.googleapis.com/skaffold/builds/latest/skaffold-linux-amd64 && \
    sudo install skaffold /usr/bin/ | bash
    ```  
*   To upgrade skaffold to a newer version, run
    ```bash
    sudo rm -f /usr/bin/skaffold && curl -Lo skaffold https://storage.googleapis.com/skaffold/builds/latest/skaffold-linux-amd64 && \
    sudo install skaffold /usr/bin/ | bash
    ```
    
#### Install `kpt`

*   Check if `kpt` is installed and the installed version is >= 0.34.0
    ```bash
    kpt version
    ```
*   If you haven't installed `kpt` previously, run
    ```bash
    sudo apt-get install google-cloud-sdk-kpt
    ```  
    
#### Install `kustomize`

*   Check if `kustomize` is installed and the installed version is >= v3.4.3
    ```bash
    kustomize version
    ```
*   If you haven't installed `kustomize` previously, run
    ```bash

>         curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"  | bash &&  sudo mv kustomize /usr/local/go/bin

Reviewer: I think we want to get rid of the extra leading whitespace here

Author: done.

    ```  
    
## Getting started

Time to complete: **About 3 minutes**

*   In your
    <walkthrough-editor-spotlight spotlightId="menu-terminal-new-terminal">terminal</walkthrough-editor-spotlight>,
    run this command to start your local minikube cluster:
    
    ```bash

>     minikube start

Reviewer: should we have a section about installing minikube in the prerequisites?

Author: cloud code auto installs minikube. https://cloud.google.com/code/docs/vscode/adding-a-cluster

        
In the example, the configurations will build an image <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="6" endLine="6" startCharacterOffset="12" endCharacterOffset="20">frontend</walkthrough-editor-select-line> and use `kpt` to *hydrate[1]* 
and deploy the applications to the cluster. 

The <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="6" endLine="6" startCharacterOffset="12" endCharacterOffset="20">frontend</walkthrough-editor-select-line> source code is stored in 
<walkthrough-editor-open-file filePath="guestbook-cl/php-redis/guestbook.php">guestbook-cl/php-redis</walkthrough-editor-open-file> 
and its configurations are stored in <walkthrough-editor-open-file filePath="guestbook-cl/config/frontend/deployment.yaml">guestbook-cl/config</walkthrough-editor-open-file>.

<walkthrough-pin-section-icon></walkthrough-pin-section-icon>
**Glossary**

> > [1] *`Hydrate`* means rendering a *kustomize* directory or a *kpt* package to a flatten 

Reviewer: ```suggestion
> [1] *`Hydrate`* means rendering a *kustomize* directory or a *kpt* package to a flattened 
```

Author: done.

    
    ```terminal
    The Deployment "frontend" is invalid: spec.template.spec.containers[0].image: Required value
    ```

*   You can add back the removed line by copying the following command and running it in the terminal.  

    ```shell script
    sed -i '18 a \        image: "frontend"'     ./config/frontend/deployment.yaml
    ```
    
<walkthrough-notification-menu-icon></walkthrough-notification-menu-icon>
**Tips**
> You can find all the `kpt` validation functions from this [catalog](https://googlecontainertools.github.io/kpt/guides/consumer/function/catalog/validators/)
> 
> Or write your own versions. See [instructions](https://googlecontainertools.github.io/kpt/guides/consumer/function/).
    
    
## Validating through a pipeline
Time to complete: **About 5 minutes**

Instead of using a single validation function, we are more interested in applying the 
configurations through a series of validation functions, each of which checks a specific rule.

To do so, you can put a list of `kpt` functions into a single file and kpt will apply the 
resources through these functions based on the function orders in the single file. 
Like the graph shows.

> ![image](https://github.com/GoogleContainerTools/skaffold/tree/master/logo/skaffold.png)

Reviewer 2: ```suggestion
![image](https://raw.githubusercontent.com/GoogleContainerTools/skaffold/master/logo/kpt-pipeline.png)
```
I believe this should be the graph since it is not referenced anywhere, and ```raw.githubusercontent.com``` should  be used, otherwise the image won't load properly.

Author: 👍

# Validate Your Configurations in skaffold Workflow

<walkthrough-disable-features toc></walkthrough-disable-features>

## Introduction

#### What is kpt?

Kpt is [an OSS tool](https://github.com/GoogleContainerTools/kpt) for Kubernetes packaging, which uses a standard format to bundle, publish, customize, update, and apply configuration manifests.

> #### What kpt can help you in skaffold

Author: Yeah, this is a duplicate content. I think it is fine because the README and tutorials are in different platform (GitHub, Cloud Shell) and it is necessary in both places to give users a basic idea what the codelab is about. Besides, we will refer the codelab in other places (e.g. kpt or other cloudshell codelab), so some users may read this content only from the cloud shell.

> #### What kpt can help you in skaffold

Reviewer: Gotcha 👍🏼

> So you don’t need to rerun the command if the file changes.

## Validating the config

Time to complete: **About 10 minutes**

Validating the configurations helps both the app development and devOps to be efficient in a 
fragile environment. 

This step uses a `kubeval` example to show how `kpt` functions[2] 
can validate the app configuration and makes the validation itself **as a declarative config.**

*   Download the kpt function resource
    
    ```bash
    kpt pkg get https://github.com/GoogleContainerTools/skaffold.git/codelab/kpt-101/resources/validation-kubeval validation-kubeval
    ```
    
*   Now let's update the skaffold.yaml to use the new validator. Replace the following code with the 
`.deploy` section in the <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="8" endLine="10" startCharacterOffset="0" endCharacterOffset="100">skaffold.yaml</walkthrough-editor-select-line>

    See the full [skaffold.yaml](https://github.com/yuwenma/sample-app/blob/kubeval/skaffold.yaml#L12) 
    ```yaml
    deploy:
      kpt:
        dir: config
        fn:
          fnPath: validation-kubeval
          network: true
    ```

> -  <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="12" endLine="12" startCharacterOffset="6" endCharacterOffset="12">.deploy.kpt.fn.fnPath</walkthrough-editor-select-line> refers to the kpt function directory we just downloaded. 

Reviewer: I wonder if we should explain a bit about how kpt functions work?

Author: Good point! I had the same doubt but my concern is that the kpt fn is not the main focus of the codelab and its concepts extends the topics too far (each codelab step is ~5 mins and it excludes hard-core concepts). So I only added the kpt fn in the *extended reading* at the end.

Reviewer: that makes sense, thanks for clarifying

    See the full [skaffold.yaml](https://github.com/yuwenma/sample-app/blob/kubeval/skaffold.yaml#L12) 
    ```yaml
    deploy:
      kpt:
        dir: config
        fn:
          fnPath: validations
          network: true
    ```
    
*   Check the result. 

    ```bash
    skaffold dev
    ```

## Verify the validation pipeline "sad path" 

Let's  break the validator pipeline. 

*   In <walkthrough-editor-open-file filePath="guestbook-cl/config/frontend/deployment.yaml>config/frontend/deployment.yaml</walkthrough-editor-open-file>, 
    remove the container's  <walkthrough-editor-select-line filePath="guestbook-cl/config/frontend/deployment.yaml" startLine="21" endLine="21" startCharacterOffset="12" endCharacterOffset="15">spec.template.spec.containers.resources.requests.cpu</walkthrough-editor-select-line> field (in line 22)  
    
    *You can **either** manually delete the line **or** run the following command*

    ```bash
    sed -i '22d' ./config/frontend/deployment.yaml
    ```

*   Check the `skaffold dev` output. 

    ```bash
    skaffold dev
    ```
    
    You should see the following warning.
    ```terminal

>     ```

Author: Good catch. Yeah, I found different outputs with the example-kubeval. I would need to double check with the fn owner and then update this field. 

Changed this to "TODO".

# Validate Your Configurations in skaffold Workflow

<walkthrough-disable-features toc></walkthrough-disable-features>

## Introduction

#### What is kpt?

Kpt is [an OSS tool](https://github.com/GoogleContainerTools/kpt) for Kubernetes packaging, which uses a standard format to bundle, publish, customize, update, and apply configuration manifests.

> #### What kpt can help you in skaffold

Reviewer 2: nit:
```suggestion
#### What kpt can help you with in skaffold
```

    You should expect the following warning about the missing field.
    
    ```terminal
    The Deployment "frontend" is invalid: spec.template.spec.containers[0].image: Required value
    ```

*   Add back the removed line and make it work again.  

    ```shell script
    sed -i '18 a \        image: "frontend"'     ./config/frontend/deployment.yaml
    ```
    
<walkthrough-notification-menu-icon></walkthrough-notification-menu-icon>
**Tips**
> You can find all the `kpt` validation functions from this [catalog](https://googlecontainertools.github.io/kpt/guides/consumer/function/catalog/validators/)
> 
> Or write your own versions. See [instructions](https://googlecontainertools.github.io/kpt/guides/consumer/function/).
    
    
## Validating through a pipeline
Time to complete: **About 5 minutes**

Instead of using a single validation function, we are more interested in applying the 
configurations through a series of validation functions, each of which checks a specific rule.

To do so, you can put a list of `kpt` functions into a single file and kpt will apply the 
resources through these functions based on the function orders in the single file. 
Like the graph shows.

> ![image](https://sites.google.com/a/google.com/d2html-img/users/yuwenma/skaffoldcodela--yvbbp5pxqzo.png)

Reviewer 2: I don't have permissions to see this image on my personal account, I believe it should be made public somehow if users are supposed to see it.

Author: Good catch. saved the image publicly.

*Let's go through an example!*    

*   Run `kpt pkg` to download the validation pipeline.
    ```bash
    kpt pkg get https://github.com/GoogleContainerTools/skaffold.git/codelab/01_kpt-validate/resources/validation-pipeline/ validations
    ```
    This pipeline contains two functions.
    *   A <walkthrough-editor-select-line filePath="guestbook-cl/validations/fn-pipeline.yaml" startLine="4" endLine="4" startCharacterOffset="8" endCharacterOffset="14">Kubeval</walkthrough-editor-select-line> validator to check the yaml schema

>     *   A <walkthrough-editor-select-line filePath="guestbook-cl/validations/fn-pipeline.yaml" startLine="15" endLine="15" startCharacterOffset="8" endCharacterOffset="20">CPUAndMemory<> validator to check if all containers have the CPU and memory reservation set.   

Reviewer 2: ```suggestion
    *   A <walkthrough-editor-select-line filePath="guestbook-cl/validations/fn-pipeline.yaml" startLine="15" endLine="15" startCharacterOffset="8" endCharacterOffset="20">CPUAndMemory</walkthrough-editor-select-line> validator to check if all containers have the CPU and memory reservation set.   
```
Missing closing tag

Author: Thanks!

*   Now let's update the <walkthrough-editor-open-file filePath="guestbook-cl/skaffold.yaml">skaffold.yaml</walkthrough-editor-open-file> 
    to use the new validator pipeline. In <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="8" endLine="13" startCharacterOffset="0" endCharacterOffset="100">skaffold.yaml</walkthrough-editor-select-line>，
    replace the `.deploy` section with the following code.

    See the full [skaffold.yaml](https://github.com/yuwenma/sample-app/blob/kubeval/skaffold.yaml#L12) 
    ```yaml
    deploy:
      kpt:
        dir: config
        fn:
          fnPath: validations
          network: true
    ```
    
*   Check the result. 

    ```bash
    skaffold dev
    ```

## Verify the validation pipeline "sad path" 

Let's  break the validator pipeline. 

> *   In <walkthrough-editor-open-file filePath="guestbook-cl/config/frontend/deployment.yaml>config/frontend/deployment.yaml</walkthrough-editor-open-file>, 

Reviewer 2: ```suggestion
*   In <walkthrough-editor-open-file filePath="guestbook-cl/config/frontend/deployment.yaml">config/frontend/deployment.yaml</walkthrough-editor-open-file>, 
```
Missing quotation

Author: done.

    remove the container's  <walkthrough-editor-select-line filePath="guestbook-cl/config/frontend/deployment.yaml" startLine="21" endLine="21" startCharacterOffset="12" endCharacterOffset="15">spec.template.spec.containers.resources.requests.cpu</walkthrough-editor-select-line> field (in line 22)  
    
    *You can **either** manually delete the line **or** run the following command*

    ```bash
    sed -i '22d' ./config/frontend/deployment.yaml
    ```

*   Check the `skaffold dev` output. 

    ```bash
    skaffold dev
    ```
    
    You should see the following warning.

    TODO: update the content.
    ```terminal

    ```

## Conclusion

<walkthrough-conclusion-trophy></walkthrough-conclusion-trophy>

> Congratulations, you've known how to validate configurations in skaffold! You can explore the full kpt configuration 

Reviewer 2: nit: 
```suggestion
Congratulations, you know how to validate configurations in skaffold! You can explore the full kpt configuration 
```

> from the skaffold.yaml[ reference doc](https://skaffold.dev/docs/references/yaml/#deploy-kpt). 

Reviewer 2: nit:
```suggestion
from the skaffold.yaml [reference doc](https://skaffold.dev/docs/references/yaml/#deploy-kpt). 
```

Author: done.

You can also try out other kpt features like `kpt pkg` and `kpt cfg` from 
[the user guide](https://googlecontainertools.github.io/kpt/reference/). They will be supported 

> in the skaffold soon. Stay tuned!  

Reviewer 2: nit:
```suggestion
in skaffold soon. Stay tuned!  
```

Author: done.
